crypto/compactcert/coinGenerator.go

```diff
@@ -0,0 +1,122 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package compactcert
+
+import (
+	"encoding/binary"
+	"golang.org/x/crypto/sha3"
+	"math/big"
+
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+// The coinChoiceSeed defines the randomness seed that will be given to an XOF function. This will be used  for choosing
+// the index of the coin to reveal as part of the compact certificate.
+type coinChoiceSeed struct {
+	// the ToBeHashed function should be updated when fields are added to this structure
+
+	partCommitment crypto.GenericDigest
+	lnProvenWeight uint64
+	sigCommitment  crypto.GenericDigest
+	signedWeight   uint64
+	data           StateProofMessageHash
+}
+
+// ToBeHashed returns a binary representation of the coinChoiceSeed structure.
+// Since this code is also implemented as a circuit in the stateproof SNARK prover we can't use
+// msgpack encoding since it may result in a variable length byte slice.
+// Alternatively, we serialize the fields in the structure in a specific format.
+func (cc *coinChoiceSeed) ToBeHashed() (protocol.HashID, []byte) {
+	signedWtAsBytes := make([]byte, 8)
+	binary.LittleEndian.PutUint64(signedWtAsBytes, cc.signedWeight)
+
+	lnProvenWtAsBytes := make([]byte, 8)
+	binary.LittleEndian.PutUint64(lnProvenWtAsBytes, cc.lnProvenWeight)
+
+	coinChoiceBytes := make([]byte, 0, len(cc.partCommitment)+len(lnProvenWtAsBytes)+len(cc.sigCommitment)+len(signedWtAsBytes)+len(cc.data))
+	coinChoiceBytes = append(coinChoiceBytes, cc.partCommitment...)
+	coinChoiceBytes = append(coinChoiceBytes, lnProvenWtAsBytes...)
+	coinChoiceBytes = append(coinChoiceBytes, cc.sigCommitment...)
+	coinChoiceBytes = append(coinChoiceBytes, signedWtAsBytes...)
+	coinChoiceBytes = append(coinChoiceBytes, cc.data[:]...)
+
+	return protocol.CompactCertCoin, coinChoiceBytes
+}
+
+// coinGenerator is used for extracting "randomized" 64 bits for coin flips
+type coinGenerator struct {
+	shkContext   sha3.ShakeHash
+	signedWeight *big.Int
```

Comment:
`signedWeight` does not need to be `big.Int`
uint64 is enough. Also, looks strange that it is defined here as big.Int, and in `coinChoiceSeed` defined as `uint64`.